Fix for tuned notes in low sampling rates

lmartorella opened the pull request:
Using integer "periods" could led to completely detuned notes in lower sampling rates.

For example, using a 8000Hz sampling rate, a period of '10' spans between 761Hz (10.5 period) to 842Hz (9.5 period).
This interval contains two notes (G5 and G5#).

In order to render better notes on lower sampling rates, perhaps a good solution is to use fixed point arithmetic for the period. Still using 16-bits, a good compromise is 12.4 (12 integer part and 4 bits of fractional).
This has enough room to render the lowest audible 20Hz on the higher 48Khz sampling rate (period of 2400 requires 12bits).

In addition, the first commit of the branch ("Fixed typo") fixes a wrong 8-bit truncation of intermediate functions `voice_wf_calc_sawtooth_period` and `voice_wf_calc_square_period`, that obviously causes wrong high pitches on longest periods.

The second commit "Added fourth octave test" could be skipped, since only introduces a test command with all the notes of the fourth octave played with low sampling rate. With this script it is possible to test the fix before and after application.

The third commit contains the fix.
No additional computational cost is used at runtime (only some int shifting). However, a 32-bit cast is required during frequency-to-period conversion. (Additional Merge Request will follow with a sequencer that 'saves' this operational cost in playback cases).

Obviously, using fixed-point periods will render uneven waveforms, creating more harmonics. This could be only reduced using over-sampling in square waves instead of using only two states (+amp and -amp).
However, given the target of the project (for MCUs) I think the result is quite OK.

Thanks, L

lmartorella committed:
- Fixed fractional frequency

lmartorella committed:
- Fixed typo

lmartorella commented:
Oh, by the way, I removed the fixed point translation in the frequency domain, since I think it less relevant for the human ears. The code was essentially the same, but applying the fp after division in my opinion is more clear. L

sjlongland commented:
Hi, I did a quick rebase of this and PR #2 onto `master`, and there was a small merge conflict with `waveform.c`… I think I managed to resolve it but haven't had a moment to test it.

https://github.com/sjlongland/atinysynth/tree/fix/fractional-freq is that composite branch, if you're happy with that we'll merge that into `master`.

Many thanks for your contributions.

lmartorella commented:
Whoops, removed one extraneous commit (the VSCode IDE settings).
For the rest, I have thought that leaving the PR as three separated commit was better for clarity, but it is OK to squash it as a single one.
I also removed the commit with the `test` file that contained the commands to build a scale (for hearing test only), to avoid conflicts.
Thx, L

sjlongland commented:
Heh, while I don't use it myself, having VS Studio Code files there isn't a big bother.  I seem to recall there was an "editorconfig" (I think that was its name) framework for a "standard" configuration file for setting up IDEs with a project -- I should look into it.

Not everyone is a vi-using luddite like me. :-)

lmartorella commented:
> Not everyone is a vi-using luddite like me. :-)
LOL, it reminds me the old times spent with DOS `edlin` :-)

However, better to not add it. VS Code is not the preferred IDE for C development. I use it for almost everything, but better to not include it in this context.

sjlongland commented:
Okay, take two… I've merged the other branch and did a rebase on this PR… https://github.com/sjlongland/atinysynth/tree/fix/fractional-freq has that implemented.  If all is well, we'll merge that.

lmartorella commented:
Thank for merging #2!
I've just rebased #3 on top of master to resolve conflicts.
Thanks, L!
